### PR TITLE
Feature/faster episode recorder

### PR DESCRIPTION
**Pull request by DavidSlayback**
# Description

This addresses two issues I've run into with RecordEpisodeStatistics

1) Silent list conversion in vectorized environments. Currently, RecordEpisodeStatistics converts terminated and truncated into lists, but obs and reward may still be arrays, and the new vectorized `info` interface uses arrays. I particularly run into this in my code if I do something like `dones.any()`. This PR leaves the underlying details of terminated/truncated alone by using NumPy array functionality.

2) Speed. Instead of looping through a potentially large batch of environments, this PR uses NumPy functions with their underlying vectorization to achieve the same task.

I considered building this with Jumpy, but I'm not sure on its status in the project, and it would also eat the cost of creating new arrays wherever I'm doing in-place modifications. 

I tested to make sure that this replicates the exact functionality of the original wrapper. To prove my speedup, I need to come up with some decent micro benchmarks; a couple attempts on Colab were fairly noisy.

## Type of change

- [x] Existing feature enhancement (non-breaking change which adds functionality)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [N/A] I have commented my code, particularly in hard-to-understand areas
- [N/A] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [N/A] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: gym/wrappers/record_episode_statistics.py [hunk 1]
```diff
@@ -114,38 +85,28 @@ def step(self, action):
...
-                else:
-                    infos = {**infos, **episode_info}
-                self.return_queue.append(episode_return)
-                self.length_queue.append(episode_length)
-                self.episode_count += 1
-                self.episode_returns[i] = 0
-                self.episode_lengths[i] = 0
+        dones = truncateds | terminateds
+        num_dones = np.sum(dones)
+        if num_dones:
```
Comment (pseudo-rnd-thoughts) [inline gym/wrappers/record_episode_statistics.py:90]: Replace with `np.any(dones)`

File: tests/wrappers/test_record_episode_statistics.py [hunk 2]
```diff
@@ -74,29 +73,3 @@ def test_wrong_wrapping_order():
 
     with pytest.raises(AssertionError):
         wrapped_env.step(wrapped_env.action_space.sample())
-
-
-def test_add_vector_episode_statistics():
```
Comment (pseudo-rnd-thoughts) [inline tests/wrappers/test_record_episode_statistics.py:79]: Why have you removed this test? 
If it fails, could you fix it or replace it with an equivalent test

File: gym/wrappers/record_episode_statistics.py [hunk 3]
```diff
@@ -8,35 +8,6 @@
 import gym
 
 
-def add_vector_episode_statistics(
```
Comment (pseudo-rnd-thoughts) [inline gym/wrappers/record_episode_statistics.py:11]: Why have you removed this function?

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Thanks for the PR, I think that the new code is nicer than the old code if it returns the same data. 
I have a couple of questions on the implement if you could ask them

File: gym/wrappers/record_episode_statistics.py (hunk 3 above)
Comment (DavidSlayback) [reply gym/wrappers/record_episode_statistics.py:11]: The function is only used for RecordEpisodeStatistics, and it undoes part of the point of this commit. 

Either way, when we finish an episode, we need to extend our `info` dictionary with the episode statistics in array form. 

In the original implementation, this was repeatedly called in a loop to produce statistics arrays and their corresponding masks, then insert them into the info dictionary

In this implementation, we create the exact same results by just using numpy indexing and the boolean done masks we already have. No need for an additional function that is used nowhere else

File: tests/wrappers/test_record_episode_statistics.py (hunk 2 above)
Comment (DavidSlayback) [reply tests/wrappers/test_record_episode_statistics.py:79]: If I remove the use of the function in RecordEpisodeStatistics, it is used nowhere else. If I delete the function, I suppose I could add a test that checks the returned EpisodeStatistics info object?

File: gym/wrappers/record_episode_statistics.py (hunk 1 above)
Comment (DavidSlayback) [reply gym/wrappers/record_episode_statistics.py:90]: I use the sum later to increment the episode count `self.episode_count += num_dones`

Comment (pseudo-rnd-thoughts) [reply gym/wrappers/record_episode_statistics.py:90]: Sorry, I didn't see that. Ignore me

File: tests/wrappers/test_record_episode_statistics.py (hunk 2 above)
Comment (pseudo-rnd-thoughts) [reply tests/wrappers/test_record_episode_statistics.py:79]: Your right, I don't think we need to replace this function with anything

File: gym/wrappers/record_episode_statistics.py (hunk 3 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/record_episode_statistics.py:11]: Ok, that makes sense

**Review by pseudo-rnd-thoughts — APPROVED** [review]
Looks good to me

**Comment by DavidSlayback** [comment]
Would it be more useful for me to move this PR over to gymnasium? Just saw that recently, didn't realize the devs were moving to a fork

**Comment by pseudo-rnd-thoughts** [comment]
Yes, we were not accepting new features after Gymnasium goes live (soon). So feel free to close here and open on gymnasium if you want

**Pull request closed without merging**